### PR TITLE
Contact history

### PR DESCRIPTION
**Pull request by ericnewcomer**
Rework the contact read page to show a more comprehensive event history. Allows toggling of non-message events and supports paging in 90 day increments. Shows the following events inline according to when the occurred:

* Campaign events
* Flow starts
* Messages
* Broadcasts
* Missed Calls
* IVR calls

Moves the contact field information to the top of the page that is expandable for nicer support for orgs with many contact fields.

Adds migration for 'purged' field on Broadcast which is in preparation for broadcasts eventually purging their message objects.

Links message status icon to channel log if there is one.

### CODE REVIEW

File: temba/contacts/views.py [hunk 1]
```diff
@@ -586,19 +581,21 @@ def contact_cmp(a, b):
...
-            context['all_messages'] = all_messages
             context['contact_sendable_urns'] = sendable_urns
             context['contact_unsendable_urns'] = unsendable_urns
-            context['contact_fields'] = ContactField.objects.filter(org=contact.org, is_active=True).order_by('pk')
-            context['contact_groups'] = contact.user_groups.extra(select={'lower_name': 'lower(name)'}).order_by('lower_name')
-            context['upcoming_events'] = contact.fire_events.filter(scheduled__gte=timezone.now()).order_by('scheduled')
+
+            # lookup all of our contact fields
+            contact_fields = ContactField.objects.filter(org=contact.org, is_active=True).order_by('label', 'pk')
+            contact_fields = [ dict(label=f.label, value=contact.get_field_display(f.key), featured=f.show_in_table) for f in contact_fields if contact.get_field_display(f.key) ]
```
Comment (rowanseymour) [inline temba/contacts/views.py]: PEP nitpick - shouldn't have spaces inside the [  ]

File: temba/utils/templatetags/temba.py [hunk 2]
```diff
@@ -23,6 +23,21 @@ def icon(o):
...
+def format_seconds(seconds):
+    if not seconds:
+        return None
+
+    if seconds < 60:
+        return '%s sec' % seconds
+    minutes = seconds / 60
+    seconds = seconds % 60
+    if seconds >= 30:
+        minutes+=1
```
Comment (rowanseymour) [inline temba/utils/templatetags/temba.py]: PEP

File: pip-requires.txt [hunk 3]
```diff
@@ -52,7 +52,7 @@ ply
 uservoice
 nose-perfdump
 geojson==1.0.7
-pycountry==1.6
```
Comment (rowanseymour) [inline pip-requires.txt]: Is there a reason we sometimes pin versions in pip-requires.txt ?

File: temba/contacts/views.py (hunk 1 above)
Comment (rowanseymour) [inline temba/contacts/views.py]: Also you could call `Contact.bulk_cache_initialize([contact])` to load all of the contact's URNs and field values in one fetch

**Comment by rowanseymour** [comment]
Really like the new look and having flow starts in there is super useful.

Should the contact field panel maybe have non-rounded corners for consistency?

**Comment by nicpottier** [comment]
Definitely love having the events in there, I have some pause around moving away from having "us" and the "contact" being on different sides of the screen. I thought that was a good step forward in having an intuitive interface when we first built that page, in that it mimicked the chat apps people used so set the right tone etc.. I feel a bit like the "flat" list is a step backwards in that regard, though obviously a step forward in showing these other events.

Any thoughts there? I'm sure you tried that and it just isn't possible to interleave the events in there and have it look nice? It just makes me a bit sad that the message list looks a lot like the old RapidSMS page. :(

**Comment by ericnewcomer** [comment]
Man, tough crowd. 

The first approach was to have a chat box approach with events tucked in similar to how the simulator works. Problems is it looks terrible on a wider screen and you really want the times lined up on all messages and events when trying to understand what has happened. Tables have their place. I think there a strong case for them here. 

We can have a chat approach like before, but I don't see a usable way that is nearly as clear as this approach even if we had two different interfaces. 

That said, definitely not interested in merging something if it feels like it's worse than before. So maybe let's just ignore this for now.

**Comment by ericnewcomer** [comment]
Ok, I think this one should pass muster now. Can I get a what what?

**Comment by rowanseymour** [comment]
:thumbsup:

**Comment by ericnewcomer** [comment]
Alright, any reservations to merging this, speak now or forever hold your peace.

**Comment by nicpottier** [comment]
go for it

**Pull request merged**